### PR TITLE
Support lazy Call.Factory initialization.

### PR DESCRIPTION
**Pull request by colinrtwhite**
- Adds `okHttpClient` and `callFactory` methods that accept a factory.
- Removed `applyCoilOptimizations` in favour of `CoilUtils. createDefaultCache`.
- This removes the code that disabled the max hosts by request limit., however I'm not sure how useful that is.

Still unsure if we should just pass `Lazy<Call.Factory>` to `ImageLoader` vs. wrapping it in a `Call.Factory`.

### CODE REVIEW

File: coil-base/src/main/java/coil/fetch/HttpUrlFetcher.kt [hunk 1]
```diff
@@ -12,7 +12,7 @@ import okhttp3.HttpUrl
 import okhttp3.Request
 
 internal class HttpUrlFetcher(
-    private val callFactory: Call.Factory
+    private val callFactory: Lazy<Call.Factory>
```
Comment (ZacSweers) [inline coil-base/src/main/java/coil/fetch/HttpUrlFetcher.kt]: Out of curiosity, why was this needed?

Comment (ZacSweers) [reply coil-base/src/main/java/coil/fetch/HttpUrlFetcher.kt]: Specifically, worried now that you'll be possibly incurring a double lazy cost now for cases like the dagger example

Comment (colinrtwhite) [reply coil-base/src/main/java/coil/fetch/HttpUrlFetcher.kt]: @ZacSweers I added support for the lazy call factory trick [here](https://github.com/coil-kt/coil/pull/50/commits/f6fd48da78d2a6478e747d28d5721bee4fd813fd#diff-a620fd560b7a610a8300f97bc2a41ea7R171), however noticed that that was creating two objects (an extra Call.Factory + the lazy delegate). I figured it would be to balance the cost between `callFactory(Call.Factory)` and `callFactory(() -> Call.Factory)`, as each one would only create one extra object. Do you think it's better to keep the `lazyCallFactory` method?

Comment (ZacSweers) [reply coil-base/src/main/java/coil/fetch/HttpUrlFetcher.kt]: I think it depends on the goal. Is it to make it lazy for a functional need or to just make the deferred initialization standard?

Comment (colinrtwhite) [reply coil-base/src/main/java/coil/fetch/HttpUrlFetcher.kt]: I'm leaning toward making deferred initialization the preferred approach since creating the disk cache can trigger strict mode warnings on some devices (https://github.com/coil-kt/coil/issues/48).

Comment (ZacSweers) [reply coil-base/src/main/java/coil/fetch/HttpUrlFetcher.kt]: Right, I think if that's the case then you have the Call.Factory API only and leave delegation to consumers to implement. If you wrap something lazily, internals shouldn't need to do about that since you can just wrap it in `Call.Factory` again. If you really want to add something as a convenience for consumers, you could do something like this:

```kotlin
inline fun ImageLoaderBuilder.callFactory(
  crossinline body: (Request) -> Call
): ImageLoaderBuilder = callFactory(object : Call.Factory {
  override fun newCall(request: Request): Call = body(request)
})

inline fun ImageLoaderBuilder.delegatingCallFactory(
  delegate: Lazy<OkHttpClient>
): ImageLoaderBuilder = callFactory {
  delegate.get().newCall(it)
}
```

Comment (colinrtwhite) [reply coil-base/src/main/java/coil/fetch/HttpUrlFetcher.kt]: Gotcha, I'll move back to using `Call.Factory` directly, however I do really like having a convenience function like:

```kotlin
fun callFactory(initializer: () -> Call.Factory) = apply {
    this.callFactory: Call.Factory = lazyCallFactory(initializer)
}
```

It's a bit more magical than `delegatingCallFactory(Lazy<OkHttpClient>)`, but I think the syntax still makes sense:

```kotlin
val imageLoader(context) {
    // Imperative
    callFactory(OkHttpClient())

    // Lazy
    callFactory { OkHttpClient() }
}
```

**Comment by hosanaSwaggart** [comment]
`<098">`
